apps/indexer/src/TzktListener.ts

```
@@ -1,214 +1,351 @@
+    const voter = await this.governanceContractIndexer.getVotingPowerForAddress(transactionEvent.sender.address, globalVotingIndex);
+    const delegates = await this.governanceContractIndexer.getDelegatesForAddress(transactionEvent.sender.address, level);
+    let delegateVotingPower = 0;
+    for (const delegate of delegates) {
+      const voter = await this.governanceContractIndexer.getVotingPowerForAddress(delegate, globalVotingIndex);
+      delegateVotingPower += voter?.votingPower ?? 0;
+    }
+
+    const winnerCandidate = transactionEvent.storage?.voting_context?.period?.promotion?.winner_candidate;
+    
+    const voteRecord: Vote = {
+      proposal_hash: winnerCandidate,
+      baker: transactionEvent.sender.address,
+      alias: transactionEvent.sender.alias,
+      contract_address: contract.address,
+      voting_power: (voter?.votingPower ?? 0) + delegateVotingPower,
+      vote: transactionEvent.parameter!.value,
+      time: transactionEvent.timestamp,
+      transaction_hash: transactionEvent.hash,
+      level
+    };
+
+    logger.info(`[TzktListener] vote @${contract.address} lvl=${level} propPeriod=${proposalPeriodIndex} promoPeriod=${promotionPeriodIndex} choice=${voteRecord.vote} vp=${voteRecord.voting_power}`);
+    await this.database.upsertVotes([voteRecord]);
+
+    const storageUrl = `${this.governanceContractIndexer.tzkt_api_url}` +
+                       `/contracts/${contract.address}/storage?level=${level}`;
+    const storageResponse = await fetch(storageUrl);
+    if (!storageResponse.ok) throw new Error(`HTTP ${storageResponse.status} fetching ${storageUrl}`);
+
+    const storageData = await storageResponse.json();
+    const promotionContext = storageData.voting_context?.period?.promotion;
+    if (!promotionContext) {
+      logger.warn(`[TzktListener] No on‑chain promotion context at lvl=${level} for ${contract.address}`);
```

We probably want to throw an error if we have a vote but the contract doesn't have a promotion